walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,229 @@
+   * determined ahead of time.
+   *
+   * @param events  The list of all events scheduled on the day of the meeting request. Not all
+   *                events must be associated with an attendee of the meeting.
+   * @param request The details of the meeting request. Duration, attendees, and optional attendees
+   *                are included here.
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    // There must be a list of events
+    if (events == null) {
+      throw new IllegalArgumentException("'events' parameter cannot be null");
+    }
+    // There must be a request
+    if (request == null) {
+      throw new IllegalArgumentException("'request' parameter cannot be null");
+    }
+    // There must be a set of attendees
+    if (request.getAttendees() == null) {
+      throw new IllegalArgumentException("'request' parameter's set of attendees cannot be null");
+    }
+    // The duration must not exceed one day
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      // returning an empty list in the case of a request that exceeds 1 day's
+      // duration is encoded in the test noOptionsForTooLongOfARequest()
+      return new LinkedList<>();
+    }
+
+    /* Initially, the entire day is available for the scheduling of the request */
+    final TimeRange WHOLE_DAY =
+            TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TimeRange.END_OF_DAY, true);
+
+    /* Find all TimeRanges during which the attendees are all available for a meeting */
+    final List<TimeRange> availableTimeRanges =
+            schedule(events, request, Arrays.asList(WHOLE_DAY), request.getAttendees());
+
+    /* Search through the TimeRanges during which the attendees are all available for a meeting,
+     * and identify the TimeRanges within those, during which optional attendees are also available for
+     * a meeting, using the same algorithm as before. */
+    final List<TimeRange> availableTimeRangesWithOptsConsidered =
+            schedule(events, request, availableTimeRanges, request.getOptionalAttendees());
+
+    /* If possible, invite optional attendees. If not, exclude them. */
+    if (!availableTimeRangesWithOptsConsidered.isEmpty()) {
+      return availableTimeRangesWithOptsConsidered;
+    } else {
+      return availableTimeRanges;
+    }
+  }
+
+  /**
+   * Schedule a meeting with the provided constraints.
+   *
+   * <p>This function runs in O(E*A) time, where E is the number of events, and A is the number of
+   * available TimeRange's.
+   *
+   * @param events              All events on the day of the request.
+   * @param request             The request made.
+   * @param availableTimeRanges The TimeRanges during which the event can be scheduled.
+   * @param attendees           The list of attendees to be invited to the event.
```

Comment:
You might want to update that this is not necessarily the list of attendees, but it can be a subset of attendees. For instance, above you use this parameter to differentiate between required and optional attendees. It would be good to document that.

Reply:
I included a comment about the design pattern used to schedule optional attendees 👍 https://github.com/thomasogara/step-repository/pull/52/commits/c87bf75a28e391e9c6815b62eed3827914ae3c86